### PR TITLE
Improve the plant hydro codes to accomodate dynamic forest

### PR DESCRIPTION
**Pull request by xuchongang**
Improve the plant hydro codes to accomodate dynamic forest
### Description:
This pull request updates the hydro codes to accomplish the following key copmonents:
1) Track the water in newly recruited plants
2) Track the water in tree mortality
3) Update the water potential with new growth of tissues and turnover of tissues

### Collaborators:
@bchristo , @ckoven , @rosiealice , @rgknox 

### Expectation of Answer Changes:
This should make the HYDRO code works for the dynamics mode. 

### Checklist:

- [X ] My change requires a change to the documentation.
- [X ] I have updated the in-code documentation .AND. (the [technical note](https://github.com/NGEET/fates-docs) .OR. the wiki) accordingly.
- [ X] I have read the [**CONTRIBUTING**](https://github.com/NGEET/fates/blob/master/CONTRIBUTING.md) document.
- [x] FATES PASS/FAIL regression tests were run
- [ ] If answers were expected to change, evaluation was performed and provided

### Test Results:

CTSM (or) E3SM (specify which) test hash-tag:

CTSM (or) E3SM (specify which) baseline hash-tag:

FATES baseline hash-tag:

Test Output:

### CODE REVIEW

**Comment by xuchongang** [comment]
@rgknox, I have pull the most recent change to avoid the recruiment

**Comment by xuchongang** [comment]
@rgknox , I think it is ready for testing and peer review.  I have also added the hydraulic failure mortality based on hydro.

File: main/FatesHistoryInterfaceMod.F90 [hunk 1]
```diff
@@ -1511,7 +1518,12 @@ subroutine update_history_dyn(this,nc,nsites,sites)
                hio_ddbh_canopy_si_scag              => this%hvars(ih_ddbh_canopy_si_scag)%r82d, &
                hio_ddbh_understory_si_scag          => this%hvars(ih_ddbh_understory_si_scag)%r82d, &
                hio_mortality_canopy_si_scag         => this%hvars(ih_mortality_canopy_si_scag)%r82d, &
-               hio_mortality_understory_si_scag     => this%hvars(ih_mortality_understory_si_scag)%r82d)
+               hio_mortality_understory_si_scag     => this%hvars(ih_mortality_understory_si_scag)%r82d, &
+               hio_h2oveg_dead_si                   => this%hvars(ih_h2oveg_dead_si)%r81d, &
```
Comment (rgknox) [inline main/FatesHistoryInterfaceMod.F90]: @xuchongang, I'm running tests on cheyenne and running into problems at this line.
The problem is that these diagnostics are only allocated if hydraulics is turned on, yet this routine wants to associate a pointer to them even when hydraulics is off, and it is returning errors.
However, I'm wondering, do we even want these diagnostics calculated in this subroutine? This is only called once per day, but my impression was that these water contents are updated sub-daily right? Shouldn't we move these calculations to update_history_hydraulics anyway?

Comment (rgknox) [reply main/FatesHistoryInterfaceMod.F90]: After looking at this again, it looks like these should be updated sub-daily:

ih_h2oveg_si
ih_h2oveg_hydro_err_si

And these should be updated daily:
ih_h2oveg_dead_si
ih_h2oveg_recruit_si
ih_h2oveg_growturn_err_si
ih_h2oveg_pheno_err_si

Seem right to you @xuchongang @bchristo ?

Comment (bchristo) [reply main/FatesHistoryInterfaceMod.F90]: @rgknox yes this is correct. thanks!

**Comment by rgknox** [comment]
Tests run on Cheyenne:
All existing tests pass:
/gpfs/fs1/scratch/rgknox/clmed-tests/fates.cheyenne.intel.dynamic-hydro-0114-v1-C295cef7-F4f0310d
Also tested b4b with master, all but nlcomp pass (expected):
/gpfs/fs1/scratch/rgknox/clmed-tests/fates.cheyenne.intel.master-C2342ece-Ffb4086b

A new exact restart test was added to CTSM for dynamic hydro.  While the "compare" phase does not pass, I think nailing down restart capabilities is more of a medium term goal. My objective in getting the restart machinery into the current PR was more about staging, and not completing the task. Will create an issue.

File: biogeophys/FatesPlantHydraulicsMod.F90 [hunk 2]
```diff
@@ -3151,14 +3883,20 @@ subroutine bisect_pv(ft, pm, lower, upper, xtol, ytol, psi_node, th_node)
     real(r8) :: y_new                  ! corresponding y value at x.new
     real(r8) :: f_new                  ! y difference between new y guess at x.new and target y
     real(r8) :: chg                    ! difference between x upper and lower bounds (approach 0 in bisection)
+    integer  :: nitr                   ! number of iterations 
     !----------------------------------------------------------------------
-  
+    if(psi_node > 0.0_r8) then
+      write(fates_log(),*)'Error: psi_note become positive,&
+                           psi_node=',psi_node
```
Comment (amametjanov) [inline biogeophys/FatesPlantHydraulicsMod.F90]: Please note that IBM compiler issues a build error on a string split into multiple lines without a continuation character on the next line: i.e. need to to update this to either
```
      write(fates_log(),*)'Error: psi_note become positive,&
                           & psi_node=',psi_node
```
or just 
```
      write(fates_log(),*)'Error: psi_note become positive, psi_node=',psi_node
```

Comment (rgknox) [reply biogeophys/FatesPlantHydraulicsMod.F90]: @amametjanov  this is something we can do and will try to enforce, thanks for bringing this up

**Pull request merged**